src/main/java/me/confor/velocity/chat/modules/GlobalChat.java

```diff
@@ -66,57 +64,61 @@ public void onPlayerChat(PlayerChatEvent event) {
             msg = msg.replaceText(config.urlReplacement);
 
         if (config.GLOBAL_CHAT_PASSTHROUGH)
-            sendMessage(msg,currentServer.get().getServer());
-        else
             sendMessage(msg);
+        else
+            sendMessage(msg, currentServer.get().getServer());
 
         if (config.GLOBAL_CHAT_TO_CONSOLE)
             this.logger.info("GLOBAL: <{}> {}", player, message);
 
-        if (!config.GLOBAL_CHAT_PASSTHROUGH)
+        /*
+        broken >=1.19.1
+
+        if (config.GLOBAL_CHAT_PASSTHROUGH)
             event.setResult(ChatResult.denied());
+         */
     }
 
     @Subscribe
-    public void onConnect(LoginEvent event) {
-        if (!config.JOIN_ENABLE)
-            return;
+    public void onConnect(ServerConnectedEvent event) {
+        RegisteredServer currentServer = event.getServer();
+        Optional<RegisteredServer> previousServer = event.getPreviousServer();
 
         String player = event.getPlayer().getUsername();
+        String server = currentServer.getServerInfo().getName();
 
-        Component msg = parseMessage(config.JOIN_FORMAT, List.of(
-                new ChatTemplate("player", player, false)
-        ));
-
-        sendMessage(msg);
-    }
-
-    @Subscribe
-    public void onDisconnect(DisconnectEvent event) {
-        if (!config.QUIT_ENABLE)
-            return;
+        Component msg;
 
-        Optional<ServerConnection> currentServer = event.getPlayer().getCurrentServer();
-        if (currentServer.isEmpty())
-            return;
+        if (previousServer.isPresent()) {
+            if (!config.SWITCH_ENABLE)
+                return;
 
-        String player = event.getPlayer().getUsername();
-        String server = currentServer.get().getServerInfo().getName();
+            msg = parseMessage(config.SWITCH_FORMAT, List.of(
+                new ChatTemplate("player", player, false),
+                new ChatTemplate("server", server, false),
+                new ChatTemplate("previous_server", previousServer.get().getServerInfo().getName(), false)
+            ));
+        } else {
+            if (!config.JOIN_ENABLE)
+                return;
 
-        Component msg = parseMessage(config.QUIT_FORMAT, List.of(
+            msg = parseMessage(config.JOIN_FORMAT, List.of(
                 new ChatTemplate("player", player, false),
                 new ChatTemplate("server", server, false)
-        ));
+            ));
 
-        if (config.GLOBAL_CHAT_PASSTHROUGH)
-            sendMessage(msg,currentServer.get().getServer());
-        else
-            sendMessage(msg);
+            if (!config.GLOBAL_CHAT_PASSTHROUGH) {
+                sendMessage(msg, currentServer);
+                return;
```

Comment:
I think I made that change because it made sense to me to group chat messages and join messages together under that toggle, as both would be duplicated in the target backend server's chat (by default) with the toggle enabled.

So, with `GLOBAL_CHAT_PASSTHROUGH` enabled, and **without cancelling chat** and disabling join messages on the backend server, players on said backend server would see both their own chat messages duplicated, as well as the join messages of players joining that server.

But yes, a separate toggle sounds like a better idea.

Reply:
A separate toggle sounds like a good idea.

Since admins might change the join/quit messages, perhaps a good idea is to work with the backend server to hide these possible duplicate messages as unilock suggests. This way the custom join/quit messages are shown network-wide but then the backend server hides its own join/quit vanilla message, so players on that one server don't see the message twice.